drivers: systick: Fix Cortex-M systick jumping forward and back again

gramsay0 opened the pull request:
The timer is started with a value of zero, after one tick it reloads to`last_load - 1`.

Check if the timer value is zero when calculating the elapsed ticks. If `COUNTFLAG` is not set and the value is zero then no ticks have occurred since the timer was started.

Add a test to validate that the issue is fixed (and stays fixed).

Extra context:
I have assumed than once `SysTick->VAL` ticks from 1 down to 0 that `COUNTFLAG` will be set and an ISR triggered.
That is why I chose to put this check in the `else` for when `COUNTFLAG` is not set.
I used `val1` rather than `val2` as `COUNTFLAG` may be set inbetween reading `SysTick->CTRL` and setting `val2`.
The check could instead be `else if (val1 == 0 && val2 == 0) {` which would allow an extra tick to be counted if the timer had reloaded in between setting `val1` and `val2`.

Input/review from anyone more familiar with this Cortex-M3 systick would be appreciated

Fixes: #48608

gramsay0 committed:
drivers: systick: Fix Cortex-M systick jumping forward and back again

The existing implementation did not properly
handle when `SysTick->VAL` is zero.

This caused three subtle edge cases:
* val1=0,COUNTFLAG=0,val2=0
  This should result in no cycles elapsed,
  however `(last_load - val2) = last_load`.
  So an extra `last_load` cycles was returned.
* val1=0,COUNTFLAG=0,val2=(last_load-1)
  This should result in 1 cycle elapsed,
  however `val1 < val2` so an extra `last_load`
  cycles was returned.
* val1=[2,1,0],COUNTFLAG=1,val2=0
  This should result in `last_load` cycles elapsed.
  However, `last_load * 2` cycles was returned.

To fix the calculation, val1 and val2 are first
wrapped/realigned from [0:last_load-1] to [1:last_load].

Tidy comments to better reflect the SysTick
behaviour and link reference manuals.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
tests: timer_monotonic: Add mps2_an385 test with icount disabled

This is to validate that GH-48608 is fixed
and stays fixed.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 commented:
I think this also fixes the networking issue seen here https://github.com/zephyrproject-rtos/zephyr/pull/63220.

I added a print statement inside the scope of the fix:

	} else if (val1 == 0) {
		/* The timer is started with a value of zero,
		 * after one tick it reloads to (last_load - 1).
		 * If the timer value is zero here, it means that no
		 * ticks have occurred since the timer was started.
		 */
		printk("Fixed\n");
		val2 = last_load;
	}

And ran tests/net/socket/tls tests on mps2_an385:

  Running TESTSUITE net_socket_tls
  ===================================================================
  START - test_close_while_accept
   PASS - test_close_while_accept in 0.022 seconds
  ===================================================================
  START - test_so_protocol
   PASS - test_so_protocol in 1.008 seconds
  ===================================================================
  START - test_so_type
   PASS - test_so_type in 1.009 seconds
  ===================================================================
  START - test_v4_dtls_sendmsg
  Fixed
   PASS - test_v4_dtls_sendmsg in 0.061 seconds
  ===================================================================
  START - test_v4_msg_trunc
  Fixed
  Fixed
   PASS - test_v4_msg_trunc in 0.026 seconds
  ===================================================================
  START - test_v4_msg_waitall
  Fixed
  Fixed
  Fixed
  Fixed
  Fixed
  Fixed
  Fixed
  Fixed
  Fixed
  Fixed
  Fixed
   PASS - test_v4_msg_waitall in 0.233 seconds
  ===================================================================
  START - test_v6_dtls_sendmsg
   PASS - test_v6_dtls_sendmsg in 0.065 seconds
  ===================================================================
  START - test_v6_msg_trunc
   PASS - test_v6_msg_trunc in 0.044 seconds
  ===================================================================
  START - test_v6_msg_waitall
   PASS - test_v6_msg_waitall in 0.510 seconds
  ===================================================================
  TESTSUITE net_socket_tls succeeded

The tests pass and you can see the print statement getting hit multiple times (`Fixed`)

gramsay0 commented:
I just realised the `|| (val1 < val2)) {` check in the preceding `if` statement may still cause this issue to occur...

microbuilder commented:
@ithinuel Want to have a look at this one? You should be able to give a proper +1 review now as well.

ithinuel commented:
If my understanding of the current implementation is correct, `elapsed()` should have something along

```c
if ((ctrl & SysTick_CTRL_COUNTFLAG_Msk) || (val1 < val2)) {
  if (!has_overflown_once) {
    /* there actually was only one clock cycle since
     * SysTick->VAL was set. Ignore this overflow. */
    has_overflown_once = true;
    val2 = last_load - 1;
  } else {
    overflow_cyc += last_load;
  }
  /* We know there was a wrap, but we might not have
   * seen it in CTRL, so clear it. */
  (void)SysTick->CTLR;
}
```

`sys_clock_driver_init` and `sys_clock_set_timeout` (both resetting `SysTick->VAL`) should also reset that global `has_overflown_once` boolean.

A thing that confuses me a bit is that in `sys_clock_driver_init` the register `SysTick->LOAD` is set to `last_load` but `sys_clock_set_timeout` sets it to `last_load - 1`. I don't understand yet the reason behind that.

gramsay0 commented:
@ithinuel

I think `COUNTFLAG` is only set when the timer ticks from 1 down to 0, not when setting `SysTick->VAL` to 0.
The implementation you propose would drop `last_load` ticks when the timer subsequently overflows (when `COUNTFLAG` is set).

> A thing that confuses me a bit is that in sys_clock_driver_init the register SysTick->LOAD is set to last_load but sys_clock_set_timeout sets it to last_load - 1. I don't understand yet the reason behind that.

I agree, maybe the `- 1` should be on line 3 rather than line 1 below:

	last_load = CYC_PER_TICK - 1;
	overflow_cyc = 0U;
	SysTick->LOAD = last_load;

gramsay0 commented:
Another option would be to just busy wait for a single cycle after setting the timer:

	SysTick->LOAD = last_load - 1;
	SysTick->VAL = 0; /* resets timer to last_load */

	while (SysTick->VAL == 0) {} /* busy wait for a single cycle */

This would avoid any complex code or race conditions, and a single cycle is pretty fast:

| Sys clock | Cycle time |
| --- | ---- |
| 10MHz | 100ns |
| 30MHz | 33ns |
| 100MHz | 10ns |

How fast this occurs is likely why it is only seen on QEMU rather than real hardware

gramsay0 commented:
Okay, I think I have a better solution now.

Instead of just clearing `COUNTFLAG` in the case of a wrap:

	/* We know there was a wrap, but we might not have
		 * seen it in CTRL, so clear it. */
		(void)SysTick->CTRL;

Reread `COUNTFLAG` to determine if the wrap was due to starting a timer or a subsequent overflow:

    (val1 < val2) && (SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk)

I have also updated some of the comments to better reflect the actual behavior of the SysTick

ithinuel commented:
> I think `COUNTFLAG` is only set when the timer ticks from 1 down to 0, not when setting `SysTick->VAL` to 0.

Indeed, I thought the flag was set on overflow/reload, but it is when it goes from 1 to 0.

> I think that would cause a ticked kernel to drop one cycle per tick

I checked the git-blame and the explanation makes sense, so I think it's correct.
`last_load` is always set to either, `MAX_CYCLES`, `MIN_DELAY` or a computed `delay`.
The `- 1` isn't required when `MAX_CYCLES` and `MIN_DELAY` are used but it's not incorrect either, the elapsed time still gets computed correctly.

ithinuel reviewed (changes requested):
I don't think reading `ctrl` for a second time is a valid solution.
The counter may have wrapped between when `val2` was last read & when the overflow actually occurred. We'd then miss some ticks.

I think the issue here is that using `val1 < val2` to detect wraps between `B` and `C` is assuming (like I did in my previous comment) that `COUNTFLAG` occurs on wrap which is wrong. I.e that `SysTick->VAL` overflow is equal to SysTick reload, which it is not.

This compounded with SysTick decrementing from 0 to 1 (wrapping at `SysTick->LOAD`) makes `SysTick->VAL` invalid to handle as is in arithmetic operations.

```c
/* if either the COUNTFLAG is set or SysTick->VAL went from 1 to
 * another value, then the counter has reloaded. */
if ((ctrl & SysTick_CTRL_COUNTFLAG_Msk)
    || ((val1 == 1) && (val2 != 1)) {
    /* There are last_load+1 cycles between each ticks. */
    overflow_cyc += last_load + 1;
    /* We know there was a wrap, but we might not have
     * seen it in CTRL, so clear it. */
    (void)SysTick->CTLR;
}

/* SysTick->VAL decrements from 0 to 1 wrapping at last_load (for
 * cycle length of last_load + 1).
 * Let's realign to a [last_load;0] count down that's easier to
 * handle in arithmetic.
 */
if (val2 == 0) {
    val2 = last_load;
} else {
    val2 -= 1;
}
return (last_load - val2) + overflow_cyc;
```

gramsay0 commented:
@ithinuel this was getting quite confusing so I just wrote out all of the possible cases from `SysTick->VAL` = -2 .. +2 (ignoring cases that could only occur on a second wrap of the counter without calling `elapsed()`).

	#include <zephyr/ztest.h>
	
	#define LAST_LOAD	  1000
	#define LL 		  LAST_LOAD
	#define INIT_OVERFLOW_CYC 1000
	
	static const struct test_case {
		uint32_t val1;
		bool countflag1;
		uint32_t val2;
		bool countflag2;
		uint32_t expected;
	} test_cases[] = {
		{ .val1 =  2,   .countflag1 = false, .val2 =  2,   .countflag2 = false, .expected = LL-2 },
		{ .val1 =  2,   .countflag1 = false, .val2 =  1,   .countflag2 = false, .expected = LL-1 },
		{ .val1 =  1,   .countflag1 = false, .val2 =  1,   .countflag2 = false, .expected = LL-1 },
		{ .val1 =  2,   .countflag1 = false, .val2 =  0,   .countflag2 = true , .expected = LL   },
		{ .val1 =  1,   .countflag1 = false, .val2 =  0,   .countflag2 = true , .expected = LL   },
		{ .val1 =  0,   .countflag1 = false, .val2 =  0,   .countflag2 = false, .expected = 0    },
		{ .val1 =  2,   .countflag1 = false, .val2 = LL-1, .countflag2 = true , .expected = LL+1 },
		{ .val1 =  1,   .countflag1 = false, .val2 = LL-1, .countflag2 = true , .expected = LL+1 },
		{ .val1 =  0,   .countflag1 = false, .val2 = LL-1, .countflag2 = false, .expected = 1    },
		{ .val1 = LL-1, .countflag1 = false, .val2 = LL-1, .countflag2 = false, .expected = 1    },
		{ .val1 =  2,   .countflag1 = false, .val2 = LL-2, .countflag2 = true , .expected = LL+2 },
		{ .val1 =  1,   .countflag1 = false, .val2 = LL-2, .countflag2 = true , .expected = LL+2 },
		{ .val1 =  0,   .countflag1 = false, .val2 = LL-2, .countflag2 = false, .expected = 2    },
		{ .val1 = LL-1, .countflag1 = false, .val2 = LL-2, .countflag2 = false, .expected = 2    },
		{ .val1 = LL-2, .countflag1 = false, .val2 = LL-2, .countflag2 = false, .expected = 2    },
		{ .val1 =  2,   .countflag1 = true , .val2 =  0,   .countflag2 = false, .expected = LL   },
		{ .val1 =  1,   .countflag1 = true , .val2 =  0,   .countflag2 = false, .expected = LL   },
		{ .val1 =  0,   .countflag1 = true , .val2 =  0,   .countflag2 = false, .expected = LL   },
		{ .val1 =  2,   .countflag1 = true , .val2 = LL-1, .countflag2 = false, .expected = LL+1 },
		{ .val1 =  1,   .countflag1 = true , .val2 = LL-1, .countflag2 = false, .expected = LL+1 },
		{ .val1 =  0,   .countflag1 = true , .val2 = LL-1, .countflag2 = false, .expected = LL+1 },
		{ .val1 = LL-1, .countflag1 = true , .val2 = LL-1, .countflag2 = false, .expected = LL+1 },
		{ .val1 =  2,   .countflag1 = true , .val2 = LL-2, .countflag2 = false, .expected = LL+2 },
		{ .val1 =  1,   .countflag1 = true , .val2 = LL-2, .countflag2 = false, .expected = LL+2 },
		{ .val1 =  0,   .countflag1 = true , .val2 = LL-2, .countflag2 = false, .expected = LL+2 },
		{ .val1 = LL-1, .countflag1 = true , .val2 = LL-2, .countflag2 = false, .expected = LL+2 },
		{ .val1 = LL-2, .countflag1 = true , .val2 = LL-2, .countflag2 = false, .expected = LL+2 },
	};
	
	static uint32_t test_elapsed(uint32_t val1, bool countflag1, uint32_t val2, uint32_t countflag2,
				     uint32_t overflow_cyc, uint32_t last_load)
	{
		if (countflag1 ||
		    ((val1 < val2 - 1) && countflag2)) {
			overflow_cyc += last_load;
		}
		if (val2 == 0) {
			return overflow_cyc;
		}
	
		return (last_load - val2) + overflow_cyc;
	}
	
	ZTEST(systick, test_elapsed) {
		for (size_t i = 0; i < ARRAY_SIZE(test_cases); i++) {
			const struct test_case *tc = &test_cases[i];
			uint32_t res = test_elapsed(tc->val1, tc->countflag1, tc->val2, tc->countflag2,
						    INIT_OVERFLOW_CYC, LAST_LOAD);
			bool pass = res - INIT_OVERFLOW_CYC == tc->expected;
			printk("%2zu %4u %u %4u %u %4u %4u %4u %s\n",
			       i, tc->val1, tc->countflag1, tc->val2,
			       tc->countflag2, res, res - INIT_OVERFLOW_CYC,
			       tc->expected, pass ? "PASS" : "FAIL");
			// zassert_true(pass);
		}
	}
	
	ZTEST_SUITE(systick, NULL, NULL, NULL, NULL, NULL);

The original code fails on these cases:

     # No change since timer start
     5    0 0    0 0 2000 1000    0 FAIL
     # Assumed COUNTFLAG would be set but it was not
     8    0 0  999 0 2001 1001    1 FAIL
    12    0 0  998 0 2002 1002    2 FAIL
    # COUNTFLAG set but VAL2 not yet wrapped, adds last_load twice
    15    2 1    0 0 3000 2000 1000 FAIL
    16    1 1    0 0 3000 2000 1000 FAIL
    17    0 1    0 0 3000 2000 1000 FAIL


My previous attempt failed on these (ignoring expected off-by-ones):

    # COUNTFLAG set but VAL2 not yet wrapped, adds last_load twice
    15    2 1    0 0 3000 2000 1000 FAIL
    16    1 1    0 0 3000 2000 1000 FAIL
    17    0 1    0 0 3000 2000 1000 FAIL

I ran your proposed solution though too (ignoring the off-by-ones):

     3    2 0    0 1 1000    0 1000 FAIL
     6    2 0  999 1 1002    2 1001 FAIL
    10    2 0  998 1 1003    3 1002 FAIL

I have updated with something that passes all these tests

---

For the other issue, `last_load` represents the total cycles for the period, `SysTick->LOAD` is set to `last_load - 1` to account for the cycle `0` --> `last_load - 1`.
So the - 1 should be on line 3 rather than line 1 below:

	last_load = CYC_PER_TICK - 1;
	overflow_cyc = 0U;
	SysTick->LOAD = last_load;

> When an interrupt then wakes the system, `isr_wrapper` calls `sys_clock_idle_exit()` which re-enables `SysTick` if `last_load` was stopped, but it does not update `last_load` to reflect the actual loading value.

That looks very suspicious....

ithinuel commented:
Oh, thank you for the test suite!
That helped me a lot wrap my head around it.

I got the current implementation to pass this suite by only "realigning" `val1` and `val2` to the range [1,last_load].
```
if (val1 == 0) { val1 = last_load; }
if (val2 == 0) { val2 = last_load; }
if (countflag || (val1 < val2)) {
    overflow_cyc += last_load;
}
return (last_load - val2) + overflow_cyc;
```

this saves us from introducing an extra read of countflag and reduces the complexity of reading the code later on.

---

>  So the - 1 should be on line 3 rather than line 1 below:

I agree

> That looks very suspicious....

Indeed, my understanding of this part of zephyr is still limited but the issue requires lots of thing to be present correctly to be triggered. Maybe it is rare enough it remained undetected?

andyross commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -101,10 +101,14 @@ static uint32_t elapsed(void)
-	 * that field.
+	/* SysTick behavior: The counter wraps after zero automatically.
+	 * The COUNTFLAG field of the CTRL register is set when it
+	 * decrements from 1 to 0. Reading the control register
```
Is that really documented behavior?  There's really a one cycle race between the wraparound and the flag bit?  (i.e. it's possible in that cycle to have a flag set but the counter read "unexpired"?).    Do emulators like qemu have the same behavior?  I bet they don't, in which case this is going to introduce bugs on the CI platforms, no?

gramsay0 replied:
>  Is that really documented behavior?
>  Do emulators like qemu have the same behavior?

Yes. A good way to understand this SysTick is to look at the policies that QEMU uses:

    s->ptimer = ptimer_init(systick_timer_tick, s,
                            PTIMER_POLICY_WRAP_AFTER_ONE_PERIOD |
                            PTIMER_POLICY_NO_COUNTER_ROUND_DOWN |
                            PTIMER_POLICY_NO_IMMEDIATE_RELOAD |
                            PTIMER_POLICY_TRIGGER_ONLY_ON_DECREMENT);

    /* Periodic timer counter stays with "0" for a one period before wrapping
     * around.  */
    #define PTIMER_POLICY_WRAP_AFTER_ONE_PERIOD (1 << 0)

    /* Make counter value of the running timer represent the actual value and
     * not the one less.  */
    #define PTIMER_POLICY_NO_COUNTER_ROUND_DOWN (1 << 4)

    /* Starting to run with/setting counter to "0" won't re-load counter
     * immediately, but after a one period.  */
    #define PTIMER_POLICY_NO_IMMEDIATE_RELOAD   (1 << 3)

    /*
     * Starting to run with a zero counter, or setting the counter to "0" via
     * ptimer_set_count() or ptimer_set_limit() will not trigger the timer
     * (though it will cause a reload). Only a counter decrement to "0"
     * will cause a trigger. Not compatible with NO_IMMEDIATE_TRIGGER;
     * ptimer_init() will assert() that you don't set both.
     */
    #define PTIMER_POLICY_TRIGGER_ONLY_ON_DECREMENT (1 << 5)
    
The last one "Only a counter decrement to "0" will cause a trigger" specifies that behavior.
(I have not dug up the actual reference manual yet... but I am pretty sure it will be in there).

ithinuel replied:
The [Armv7-m Architecture Reference Manual](https://developer.arm.com/documentation/ddi0403/ee/?lang=en) (chapter B3.3.1) says:
> When the counter transitions to zero, it sets the COUNTFLAG status bit to 1. Reading the COUNTFLAG status bit
clears it to 0.
>
>Writing to SYST_CVR clears both the register and the COUNTFLAG status bit to zero. This causes the SysTick logic to reload SYST_CVR from SYST_RVR on the next timer clock. A write to SYST_CVR does not trigger the SysTick exception logic.
>
> Reading SYST_CVR returns the value of the counter at the time the register is accessed.

where `SYST_CVR` is `SysTick->VAL` and `SYST_RVR` is `SysTick->LOAD`.

Armv6-m & Armv8-m are sensibly the same.
See:
- [Armv6-m Architecture Reference Manual](https://developer.arm.com/documentation/ddi0419/e/?lang=en) chapter B3.3.1
- [Armv8-m Architecture Reference Manual](https://developer.arm.com/documentation/ddi0553/bw/?lang=en) chapter B11.1

gramsay0 commented:
@ithinuel that is a very nice way to fix this issue, let's go with that, thanks!

It also shows that the original author had it mostly correct, just a subtle off-by-one misunderstanding of how the counter functions

andyross commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -115,6 +119,13 @@ static uint32_t elapsed(void)
+	}
+	if (val2 == 0) {
+		val2 = last_load;
+	}
```
I guess I'm having trouble understanding the logic here?  Seems like this is going to implement a N-1 cycle wraparound, which will lose time?  Or maybe that's OK with systick as the hardware is racy anyway.

I guess my gut says that what you want to do here if you detect you're in the 1-cycle race is just to retry the operation in the "safe" zone, no?

gramsay0 replied:
No cycles are lost. The naming of `last_load` is slightly confusing as it is the full N cycles, the timer is actually (re)loaded with `last_load - 1` cycles.

> I guess my gut says that what you want to do here if you detect you're in the 1-cycle race is just to retry the operation in the "safe" zone, no?

That would be another option, but it seems unnecessary given this simple fix

andyross reviewed (commented):
Some notes, not a -1 as I know this is a weird device and I'm not expert enough.  I do worry though that what this is treating might be better described as a "hardware bug" and won't apply to all systick platforms symmetrically?

ithinuel commented:
Since this code relies on the fact that `SysTick->LOAD = last_load - 1`, I am wondering if this PR shouldn't also include the fix [line 303](https://github.com/zephyrproject-rtos/zephyr/pull/63344/files#diff-7136b8e56a78f89de9d57f25e8de1296fc20155b63d7876c3258a7d0b7245325L303)?

gramsay0 committed:
drivers: systick: Fix Cortex-M SysTick dropping 1 cycle per tick

`last_load` is the full N cycles and `SysTick->LOAD` should
be loaded with `last_load - 1` for the calculations work
correctly.

Note: This only affects a kernel in ticked operation.
Tickless kernels periodically restart the timer correctly.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 commented:
> Since this code relies on the fact that `SysTick->LOAD = last_load - 1`, I am wondering if this PR shouldn't also include the fix [line 303](https://github.com/zephyrproject-rtos/zephyr/pull/63344/files#diff-7136b8e56a78f89de9d57f25e8de1296fc20155b63d7876c3258a7d0b7245325L303)?

Ok, added that fix too

andyross reviewed (dismissed):
I'm sold, thanks for the doc reference.  It wouldn't hurt to put a link to that in the source, FWIW, as that race is very surprising.

gramsay0 commented:
@andyross ok, done. I also added this comment to clarify what the change actually does:

	 First, manually wrap/realign val1 and val2 from [0:last_load-1]
	 to [1:last_load]. This allows subsequent code to assume that
	 COUNTFLAG and wrapping occur on the same cycle.
